### ISSUE TITLE
Tensor forward output drifts on round-trip after in-place training despite identical params (Clone-after-Train non-determinism)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#311]
## Summary

A neural network's `Predict` output drifts by ~3-4% after a serialize/deserialize round-trip, even though all layer parameters are byte-identical between the original (post-training) and the cloned model. The drift is reproducible with parameter-level diff = 0 but observable forward-output diff.

This violates the paper-implied contract that a saved model reloaded should reproduce its exact predictions (Salakhutdinov & Hinton 2009 implicitly assumes this for their Deep Boltzmann Machine experiments — it's a fundamental property of any production-grade model serialization).

## Reproduction

```csharp
using AiDotNet.NeuralNetworks;
using AiDotNet.Tensors.LinearAlgebra;

var network = new DeepBoltzmannMachine<double>();  // default 128-dim DBM
var trainInput = new Tensor<double>(new[] { 128 });
var trainTarget = new Tensor<double>(new[] { 128 });
var rng = new System.Random(42);
for (int i = 0; i < trainInput.Length; i++) trainInput[i] = rng.NextDouble() * 2 - 1;
for (int i = 0; i < trainTarget.Length; i++) trainTarget[i] = rng.NextDouble() * 2 - 1;

for (int i = 0; i < 10; i++) network.Train(trainInput, trainTarget);
network.SetTrainingMode(false);

var probe = new Tensor<double>(new[] { 128 });
for (int i = 0; i < probe.Length; i++) probe[i] = rng.NextDouble() * 2 - 1;

var trainedOut = network.Predict(probe);

var cloned = (DeepBoltzmannMachine<double>)network.Clone();
cloned.SetTrainingMode(false);
var clonedOut = cloned.Predict(probe);

// Per-layer parameter compare:
// Layer 0 RBMLayer:  orig=64628 cloned=64628 maxDelta=0.000E+000
// Layer 1 RBMLayer:  orig=251000 cloned=251000 maxDelta=0.000E+000
// Layer 2 DenseLayer: orig=64128 cloned=64128 maxDelta=0.000E+000

// Output compare:
// trainedOut[3] = 0.0656, clonedOut[3] = 0.0590  ← 0.0066 diff
// trainedOut[34] = 0.4651, clonedOut[34] = 0.4600 ← 0.0051 diff
// max(|trainedOut - clonedOut|) ≈ 4e-2 absolute, ≈ 4% relative
```

## Self-determinism check passes

`network.Predict(probe)` called twice on the same network produces byte-identical output (`maxDiff = 0.000E+000`).

## Clone-of-clone is stable

```csharp
var cloned2 = (DeepBoltzmannMachine<double>)cloned.Clone();
var cloned2Out = cloned2.Predict(probe);
// max(|clonedOut - cloned2Out|) = 0.000E+000  ← stable across further round-trips
// max(|trainedOut - clonedOut|)  = 3.935E-002 ← drift only on FIRST round-trip
```

So the bug is one-shot: once `SetParameters` has put the tensor data into a canonical layout, further round-trips are deterministic. The drift is between (a) the in-place updated tensor produced by `Engine.TensorSubtract(weights, scaledGrad)` during training and (b) the freshly-allocated tensor produced by `SetParameters` during deserialize.

## Root cause hypothesis

`Engine.TensorSubtract`, `Engine.TensorMultiplyScalar`, `Engine.TensorMatMul`, etc. allocate output tensors from a SIMD-padded buffer pool. The returned `Tensor<T>` carries the logical shape, but the underlying `Data` buffer has alignment-padded extra capacity. On subsequent `Engine.TensorMatMul` / `Engine.TensorTranspose` operations during forward, the SIMD reduction order across that padded buffer differs from the reduction order across a freshly-allocated, non-padded buffer (which is what `Tensor.SetFlat` produces in `RBMLayer.SetParameters`).

Float64 reduction order is non-associative, so even with logically-identical input values, the rounding sequence diverges at the third or fourth significant digit, accumulating to ~4% error after passing through 3 layers of `MatMul + sigmoid`.

## Expected behavior

`network.Clone()` followed by `cloned.Predict(x)` should produce **bit-identical** output to `original.Predict(x)` — this is the standard PyTorch / TensorFlow contract for `model.state_dict()` round-trips.

Either:
- SIMD-pool-allocated tensors should normalize their layout before being read by downstream ops, or
- `TensorMatMul` / `TensorTranspose` should be layout-invariant (e.g., enforce a canonical reduction order regardless of buffer padding).

## Tested versions

AiDotNet.Tensors version currently referenced by ooples/AiDotNet master (see `Directory.Packages.props`).

## Why it matters for AiDotNet

The downstream PR ooples/AiDotNet#1224 cluster F has `DeepBoltzmannMachineTests.Clone_AfterTraining_ShouldPreserveLearnedWeights` blocked on this:

```
||Δ|| = 1.806E-001, tolerance = 6.777E-005, ||trained|| = 6.777E+000
```

Other models with shorter forward chains (single Conv + activation, etc.) drift below the test's tolerance and pass; DBM's 3-layer 128 → 500 → 500 → 128 chain accumulates the drift past the threshold.

The paper-faithful fix is a deterministic forward, regardless of how the tensor was produced — that's a clean engine-level invariant rather than a per-model workaround.

**Comment by ooples** [comment]
## Confirmed still reproducing on Tensors 0.75.2

After bumping AiDotNet.Tensors from 0.72.0 → 0.75.2 in the consumer (ooples/AiDotNet PR #1274), the bug is still 100% reproducible. Tightened diagnostic via `tools/DbmCloneDebug`:

```
Self-determinism: original max-diff=0.000E+000, cloned max-diff=0.000E+000
Param count: 379756, max param diff: 0.000E+000
Out diff: max abs = 3.595E-002
  out[0]: trained=0.07058457  cloned=0.06588474  diff=4.700E-003
  out[1]: trained=0.13531564  cloned=0.12300100  diff=1.231E-002
  out[2]: trained=0.89391353  cloned=0.89220530  diff=1.708E-003
  out[3]: trained=0.11805855  cloned=0.09551822  diff=2.254E-002
  out[4]: trained=0.10255315  cloned=0.08972200  diff=1.283E-002
```

What this tells us:
- **Self-determinism is perfect** on both networks (consecutive `Predict(x)` calls bit-identical → ops are deterministic functions of their tensor operands).
- **Parameters are byte-identical** between original and cloned (`max abs param diff = 0.0`, all 379,756 doubles).
- **Forward output diverges by 3.6% absolute** despite identical parameters.

Since the ops are deterministic AND the parameter VALUES are identical, the divergence has to come from something the forward kernel reads beyond the logical buffer (alignment padding, allocator-pool tag, capacity-vs-length mismatch, etc.) — i.e. a difference in the underlying `Tensor<T>` instance state that isn't captured by `ToArray()` / `ToVector()`.

## User-side workaround attempts that did NOT help

I tried two user-side fixes in PR #1274 to canonicalize the cloned-side `_weights` buffer:

1. **DBM `DeserializeNetworkSpecificData`**: pushing each `_layerWeights[i]` through `Engine.TensorMultiplyScalar(t, 1.0)` after deserialize — drift unchanged (1.96E-001 → 1.96E-001). Note this targets a dead-code path; supervised DBM uses `Layers` chain, not `_layerWeights`.

2. **`RBMLayer.SetParameters`**: same `TensorMultiplyScalar(t, 1.0)` round-trip on `_weights`, `_visibleBiases`, `_hiddenBiases` after `SetFlat` writes — drift unchanged (1.81E-001).

Either `TensorMultiplyScalar(x, 1.0)` is short-circuited to identity in 0.75.2 (returns the same tensor without re-allocating), or the layout difference isn't what TensorMultiplyScalar's output path produces either. **No load-bearing user-side workaround appears to exist** — the fix has to be on the engine.

## Suggested fix direction (user perspective)

If the divergence is alignment-padding / pool-tag related, the engine-side fix is to enforce a canonical layout invariant: any `Tensor<T>` that flows into `FusedLinear` / `TensorMatMul` / `TensorTranspose` must be in a single canonical buffer state regardless of how it was constructed (engine-op output vs. fresh `new Tensor<T>(shape)` vs. `SetFlat` populated).

A weaker fix: a public `Engine.Canonicalize(Tensor<T>)` method that consumers can explicitly call after `SetFlat`-style population to guarantee subsequent kernels produce the same output as if the tensor came from an engine op. PyTorch's `tensor.contiguous()` is the analog.

PR #1274 is keeping B14 (`DeepBoltzmannMachineTests.Clone_AfterTraining_ShouldPreserveLearnedWeights`) marked as blocked-on-this-issue until a user-callable canonicalization API or a kernel-side fix ships.

**Comment by ooples** [comment]
## v0.75.3 verification: bug still reproduces

Bumped consumer (`ooples/AiDotNet` PR #1274) to AiDotNet.Tensors 0.75.3 and re-ran the `tools/DbmCloneDebug` diagnostic from [comment 4407032500](https://github.com/ooples/AiDotNet.Tensors/issues/311#issuecomment-4407032500). Result:

```
Tensors 0.75.3 diagnostic:
  Param count: 379756, max param diff: 0.000E+000   ← byte-identical params
  Out diff: max abs = 3.508E-002                    ← 3.5% output drift, unchanged
    out[0]: trained=0.10870797  cloned=0.10517597  diff=3.532E-003
    out[1]: trained=0.07304232  cloned=0.06283829  diff=1.020E-002
    out[2]: trained=0.91160650  cloned=0.91561788  diff=4.011E-003
    out[3]: trained=0.27038165  cloned=0.26825793  diff=2.124E-003
    out[4]: trained=0.12109474  cloned=0.09941344  diff=2.168E-002
```

Drift magnitude across versions:
- 0.72.0: ~3.6 % output diff
- 0.75.2: ~3.6 % output diff (unchanged)
- **0.75.3: ~3.5 % output diff (unchanged after #315 zero-pad fix)**

Self-determinism remains perfect on both networks. Parameters are still byte-identical between original and cloned. Output still diverges by ~3.5 %.

The #315 fix ("zero pooled-tensor padding so SIMD overhang reads can't leak garbage") addresses one slice of the pooled-buffer-state space, but apparently not the slice that's affecting `Engine.TensorMatMul` / `Engine.FusedLinear` / `Engine.TensorTranspose` reductions on the post-Train tensors. The empirical signature still matches the prior root-cause hypothesis: forward kernels read past the logical extent into pool-allocated state that differs between training-produced tensors and SetParameters-populated tensors.

Could you reopen / re-investigate? The consumer test [`DeepBoltzmannMachineTests.Clone_AfterTraining_ShouldPreserveLearnedWeights`](https://github.com/ooples/AiDotNet/blob/master/tests/AiDotNet.Tests/ModelFamilyTests/Base/NeuralNetworkModelTestBase.cs) still fails with `||Δ|| = 1.651E-001 > tolerance 6.770E-005` on 0.75.3.

If a debugger / heap snapshot would help, I can attach the failing process state — let me know what artifacts are most useful.
